### PR TITLE
Add Vulkan validation layers to GAPID APK

### PR DESCRIPTION
**Pull request by hevrard**
We need to have access to the ANDROID_NDK_HOME environment variable,
hence the new repository_rule.

Google bug: b/141360600

### CODE REVIEW

**Comment by hevrard** [comment]
@pmuetschard this is my first meaningful contribution in Bazel, not sure whether I'm doing it right or not, feedback very welcome!

I'm still facing Bazel warnings like the following:

```
WARNING: /path/to/work/gapid/gapidapk/android/apk/BUILD.bazel:31:1:
in srcs attribute of cc_library rule //gapidapk/android/apk:armeabi-v7a_native:
please do not import '@android_native_vulkan_validation_layers//:armeabi-v7a/libVkLayer_threading.so'
directly. You should either move the file to this package or depend on an appropriate
rule there. Since this rule was created by the macro 'gapid_apk', the error might have
been caused by the macro implementation in /path/to/work/gapid/gapidapk/android/apk/rules.bzl:102:16
```

Do you have any idea what needs hot to properly get rid of these?

**Comment by AWoloszyn** [comment]
Does the NDK not get pulled in as a part of the build?
Can we pull the NDK in as part of the build itself?

**Comment by AWoloszyn** [comment]
Looks like the ndk already comes in like this:

```
        maybe_repository(
            native.android_ndk_repository,
            name = "androidndk",
            locals = locals,
            api_level = 23, # This is the minimum API
        )
```

IT would be good if we can just point to the existing repository, rather than needing the ANDROID_NDK_HOME as well?

File: tools/build/rules/android.bzl [hunk 1]
```diff
@@ -105,3 +105,35 @@ android_native_app_glue = repository_rule(
         "ANDROID_NDK_HOME",
     ]
 )
+
+# Retrieve Vulkan validation layers from the Android NDK
+
+def _android_native_vulkan_validation_layers(ctx):
+    build = ""
+    for abi in ["armeabi-v7a", "arm64-v8a", "x86", "x86_64"]:
```
Comment (AWoloszyn) [inline tools/build/rules/android.bzl]: We don't ship x86_64 for Android.

File: tools/build/rules/android.bzl [hunk 2]
```diff
@@ -105,3 +105,35 @@ android_native_app_glue = repository_rule(
...
+
+        for layer in ["core_validation", "object_tracker", "parameter_validation", "threading", "unique_objects"]:
```
Comment (AWoloszyn) [inline tools/build/rules/android.bzl]: If we update the NDK version (if we can), these have been renamed to be a single layer, might be worth updating GAPID for this.

File: tools/build/rules/android.bzl [hunk 3]
```diff
@@ -105,3 +105,35 @@ android_native_app_glue = repository_rule(
...
+            layerpath = abi + "/libVkLayer_" + layer + ".so"
+            ctx.symlink(
```
Comment (AWoloszyn) [inline tools/build/rules/android.bzl]: Does this work on Windows?

File: tools/build/rules/android.bzl (hunk 2 above)
Comment (hevrard) [reply tools/build/rules/android.bzl]: The unified layer ships in NDK r21, which is still in Beta as of today (https://developer.android.com/ndk/downloads)

**Comment by hevrard** [comment]
The Android NDK Bazel rules are auto-generated by Bazel itself, and they offer a predefined set of rules. The BUILD file for the NDK can be found in `~/.cache/bazel/.../external/androidndk/BUILD.bazel`

I can't see an easy way to refer to the validation layers using the predefined NDK rules / targets. I opened https://github.com/bazelbuild/bazel/issues/10094 to ask for new filegroups to easily access the Validation layers.

File: tools/build/rules/android.bzl (hunk 3 above)
Comment (hevrard) [reply tools/build/rules/android.bzl]: Yes, confirmed by the CI build, I manually checked that the validation layers are present in the generated APKs

File: tools/build/workspace.bzl [hunk 4]
```diff
@@ -278,6 +278,12 @@ def gapid_dependencies(android = True, mingw = True, locals = {}):
             locals = locals,
         )
 
+        maybe_repository(
+            android_native_vulkan_validation_layers,
```
Comment (pmuetschard) [inline tools/build/workspace.bzl]: I would just call the rule and the repository `ndk_vk_validation_layers`

File: tools/build/rules/android.bzl [hunk 5]
```diff
@@ -105,3 +105,35 @@ android_native_app_glue = repository_rule(
...
+                ctx.path(ctx.os.environ["ANDROID_NDK_HOME"] +
+                         "/sources/third_party/vulkan/src/build-android/jniLibs/" + layerpath),
+                ctx.path(layerpath),
+            )
+
+        build += "\n".join([
+            "filegroup(",
+            "    name = \"vulkan_validation_layer_" + abi + "\",",
```
Comment (pmuetschard) [inline tools/build/rules/android.bzl]: Just use the abi as the rule name. The `vulkan_validation_layer` part is redundant with the workspace name.

File: tools/build/workspace.bzl (hunk 4 above)
Comment (hevrard) [reply tools/build/workspace.bzl]: done

File: tools/build/rules/android.bzl (hunk 5 above)
Comment (hevrard) [reply tools/build/rules/android.bzl]: done

File: tools/build/rules.bzl [hunk 6]
```diff
@@ -16,6 +16,7 @@ load("//tools/build/rules:android.bzl",
     _android_native_app_glue = "android_native_app_glue",
     _android_native = "android_native",
     _android_native_binary = "android_native_binary",
+    _android_native_vulkan_validation_layers = "android_native_vulkan_validation_layers",
```
[the remark this answers is not in the document]
Comment (hevrard) [reply tools/build/rules.bzl]: I was following the native app glue indeed! Fixed.

File: gapidapk/android/apk/rules.bzl [hunk 7]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-load("//tools/build:rules.bzl", "extract", "filehash")
+load("//tools/build:rules.bzl", "android_native_vulkan_validation_layers", "extract", "filehash")
```
[the remark this answers is not in the document]
Comment (hevrard) [reply gapidapk/android/apk/rules.bzl]: removed

File: gapidapk/android/apk/rules.bzl [hunk 8]
```diff
@@ -102,7 +102,11 @@ def gapid_apk(name = "", abi = "", pkg = "", libs = {}, bins = {}):
         srcs = select({
             "//tools/build:android-" + name: [name + "_libs"],
             "//conditions:default": [],
-        })
+        }),
+        deps = select({
+            "//tools/build:android-" + name: ["@ndk_vk_validation_layer//:" + abi],
+            "//conditions:default": [],
+        }),
```
Comment (pmuetschard) [inline gapidapk/android/apk/rules.bzl]: does it not work if you just make it a dep below in the android_binary target?

Comment (hevrard) [reply gapidapk/android/apk/rules.bzl]: I tried this in native.android_binary:

```
        deps = [
            "//gapidapk/android/app/src/main:gapid",
            ":" + name + "_native",
        ] + select({
            "//tools/build:android-" + name: ["@ndk_vk_validation_layer//:" + abi],
        },
	no_match_error = "ERROR"
	),
```

This does hit the error, I'm not sure why we cannot rely on the same condition as in native.android_library?

Another option is:

```
        deps = [
            "//gapidapk/android/app/src/main:gapid",
            ":" + name + "_native",
            "@ndk_vk_validation_layer//:" + abi
        ],
```

But this leads to all APKs embedding validation layers for all ABIs.

**Pull request merged**